src/utils/OnboardingProvider.tsx

```diff
@@ -0,0 +1,351 @@
+'use client';
+
+import {
+  Profile,
+  ProfileLanguage,
+  ProfileRole,
+  RoleEnum,
+} from '@/types/schema';
+import {
+  createContext,
+  useState,
+  useMemo,
+  ReactNode,
+  SetStateAction,
+  Dispatch,
+  useContext,
+} from 'react';
+import { UUID } from 'crypto';
+import { ProfileContext } from './ProfileProvider';
+
+interface FlowData {
+  url: string;
+  name: string;
+}
+
+interface OnboardingContextType {
+  profile: Profile;
+  canReads: Set<string>;
+  canWrites: Set<string>;
+  roles: Set<RoleEnum>;
+  progress: number;
+  flow: FlowData[];
+  canContinue: boolean;
+  updateProfile: (updateInfo: Partial<Profile>) => void;
+  flushData: () => Promise<void>;
+  setProgress: Dispatch<SetStateAction<number>>;
+  setCanContinue: Dispatch<SetStateAction<boolean>>;
+  setFlow: Dispatch<SetStateAction<FlowData[]>>;
+  setCanReads: Dispatch<SetStateAction<Set<string>>>;
+  setCanWrites: Dispatch<SetStateAction<Set<string>>>;
+  setRoles: Dispatch<SetStateAction<Set<RoleEnum>>>;
+}
+
+export const OnboardingContext = createContext<
+  OnboardingContextType | undefined
+>(undefined);
+
+const blankProfile: Profile = {
+  first_name: '',
+  last_name: '',
+  hours_per_month: 0,
+  location: '',
+  start_date: '',
+  user_id: '0-0-0-0-0',
+};
+
+export default function OnboardingProvider({
+  children,
+}: {
+  children: ReactNode;
+}) {
+  const profileCtx = useContext(ProfileContext);
+  const [progress, setProgress] = useState(0);
+  const [flow, setFlow] = useState<FlowData[]>([]);
+  const [profile, setProfile] = useState<Profile>({ ...blankProfile });
+  const [canReads, setCanReads] = useState<Set<string>>(new Set());
+  const [canWrites, setCanWrites] = useState<Set<string>>(new Set());
+  const [roles, setRoles] = useState<Set<RoleEnum>>(new Set());
+  const [canContinue, setCanContinue] = useState<boolean>(false);
+
+  const providerValue = useMemo(() => {
+    /**
+     * Updates stored profile state with the partial data.
+     * Does not affect database.
+     */
+    const updateProfile = (updatedInfo: Partial<Profile>) => {
+      const newProfileData = { ...profile, ...updatedInfo };
+      setProfile(newProfileData);
+    };
+
+    const flushData = async () => {
```

Comment:
clarifying question: do we call `flushData` at the end of the onboarding flow? (sorry im having some trouble keeping track of all the different operations lol)

Reply:
yis